Build fails with “'CGWindowListCreateImage' is unavailable: obsoleted in macOS 15.0”

eric-carlson opened the pull request:
#### ea20b8a718cd18f88b5ba0fd7536c9f13c55f89e
<pre>
Build fails with “&apos;CGWindowListCreateImage&apos; is unavailable: obsoleted in macOS 15.0”
<a href="https://bugs.webkit.org/show_bug.cgi?id=277564">https://bugs.webkit.org/show_bug.cgi?id=277564</a>
<a href="https://rdar.apple.com/133237307">rdar://133237307</a>

Reviewed by Jean-Yves Avenard.

Softlink `CGWindowListCreateImage` so we can call it in PAL to avoid the obsolete error
when building. Ultimately we should replace it by using ScreenCaptureKit, but doing so
causes a 2% Speedometer regression so do this for now.

* Source/WebCore/PAL/pal/cg/CoreGraphicsSoftLink.cpp:
* Source/WebCore/PAL/pal/cg/CoreGraphicsSoftLink.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cg/CGWindowUtilities.cpp: Added.
(WebCore::cgWindowListCreateImage):
* Source/WebCore/platform/graphics/cg/CGWindowUtilities.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _windowSnapshotInRect:withOptions:]):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(takeWindowSnapshot):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::takeWindowSnapshot):
* Source/WebKitLegacy/mac/WebView/WebFullScreenController.mm:
(-[WebFullScreenController enterFullScreen:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _windowSnapshotInRect:withOptions:]):
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h:
* Tools/DumpRenderTree/mac/PixelDumpSupportMac.mm:
(takeWindowSnapshot):
* Tools/MiniBrowser/mac/BrowserWindowController.h:
* Tools/MiniBrowser/mac/BrowserWindowController.m:
(-[BrowserWindowController windowSnapshotInRect:]):
(-[BrowserWindowController sharingService:transitionImageForShareItem:contentRect:]):
* Tools/MiniBrowser/mac/WK1BrowserWindowController.m:
(-[WK1BrowserWindowController windowSnapshotInRect:]):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController windowSnapshotInRect:]):
* Tools/WebKitTestRunner/mac/PlatformWebViewMac.mm:
(WTR::PlatformWebView::windowSnapshotImage):

Canonical link: <a href="https://commits.webkit.org/282791@main">https://commits.webkit.org/282791@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25a099967dd5c6034fec582ae4c133cec7d33c84

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64261 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43618 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16858 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68283 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14869 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66381 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51316 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15149 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51715 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10252 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67330 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40327 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55596 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32334 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36997 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12978 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13743 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58959 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13307 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69982 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8208 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12828 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59036 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8241 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55691 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59197 "Found 2 new API test failures: /WebKitGTK/TestContextMenu:/webkit/WebKitWebView/popup-event-signal, /WebKitGTK/TestContextMenu:/webkit/WebKitWebView/context-menu-key (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14189 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6783 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/467 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39438 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40517 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41700 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40260 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/graphics/cg/CGWindowUtilities.h:
```diff
@@ -0,0 +1,38 @@
+#include <CoreGraphics/CoreGraphics.h>
+#include <math.h>
+
+namespace WebCore {
```
should be within `#if USE(CG)`

jyavenard commented on Source/WebCore/platform/graphics/cg/CGWindowUtilities.cpp:
```diff
@@ -0,0 +1,45 @@
+
+namespace WebCore {
+
+CGImageRef cgWindowListCreateImage(CGRect screenBounds, CGWindowListOption listOption, CGWindowID windowID, CGWindowImageOption imageOption)
```
Could we return a RetainPtr<CGImageRef> here considering it's no longer a direct call to CoreGraphic?

eric-carlson replied:
This is meant as a temporary change as we should use ScreenCapture Kit but can’t until we figure out the perf regression the previous patch caused, so I would rather not have to change all of the call sites

jyavenard commented on Source/WebCore/PAL/pal/cg/CoreGraphicsSoftLink.cpp:
```diff
@@ -39,4 +39,6 @@ SOFT_LINK_FRAMEWORK_FOR_SOURCE_WITH_EXPORT(PAL, CoreGraphics, PAL_EXPORT)
 SOFT_LINK_FUNCTION_FOR_SOURCE_WITH_EXPORT(PAL, CoreGraphics, CGContextSetOwnerIdentity, void, (CGContextRef context, task_id_token_t owner), (context, owner), PAL_EXPORT)
 #endif
 
+SOFT_LINK_FUNCTION_MAY_FAIL_FOR_SOURCE_WITH_EXPORT(PAL, CoreGraphics, CGWindowListCreateImage, CGImageRef, (CGRect screenBounds, CGWindowListOption listOption, CGWindowID windowID, CGWindowImageOption imageOption), (screenBounds, listOption, windowID, imageOption), PAL_EXPORT)
```
shouldn't this be defined only on Mac? 

Seems to be a Mac only API
https://developer.apple.com/documentation/coregraphics/1454852-cgwindowlistcreateimage

eric-carlson replied:
Oh, you mean it should actually compile on other ports ;-)

jyavenard reviewed (approved):
r+ with comments

jyavenard commented on Source/WebCore/platform/graphics/cg/CGWindowUtilities.cpp:
```diff
@@ -0,0 +1,45 @@
+#include "config.h"
+#include "CGWindowUtilities.h"
+
+#if USE(CG)
```
I think this should be `#IF USE(CG) && PLATFORM(MAC)`

eric-carlson committed:
Build fails with “'CGWindowListCreateImage' is unavailable: obsoleted in macOS 15.0”
https://bugs.webkit.org/show_bug.cgi?id=277564
rdar://133237307

Reviewed by Jean-Yves Avenard.

Softlink `CGWindowListCreateImage` so we can call it in PAL to avoid the obsolete error
when building. Ultimately we should replace it by using ScreenCaptureKit, but doing so
causes a 2% Speedometer regression so do this for now.

* Source/WebCore/PAL/pal/cg/CoreGraphicsSoftLink.cpp:
* Source/WebCore/PAL/pal/cg/CoreGraphicsSoftLink.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/cg/CGWindowUtilities.cpp: Added.
(WebCore::cgWindowListCreateImage):
* Source/WebCore/platform/graphics/cg/CGWindowUtilities.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _windowSnapshotInRect:withOptions:]):
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(takeWindowSnapshot):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::takeWindowSnapshot):
* Source/WebKitLegacy/mac/WebView/WebFullScreenController.mm:
(-[WebFullScreenController enterFullScreen:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _windowSnapshotInRect:withOptions:]):
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h:
* Tools/DumpRenderTree/mac/PixelDumpSupportMac.mm:
(takeWindowSnapshot):
* Tools/MiniBrowser/mac/BrowserWindowController.h:
* Tools/MiniBrowser/mac/BrowserWindowController.m:
(-[BrowserWindowController windowSnapshotInRect:]):
(-[BrowserWindowController sharingService:transitionImageForShareItem:contentRect:]):
* Tools/MiniBrowser/mac/WK1BrowserWindowController.m:
(-[WK1BrowserWindowController windowSnapshotInRect:]):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController windowSnapshotInRect:]):
* Tools/WebKitTestRunner/mac/PlatformWebViewMac.mm:
(WTR::PlatformWebView::windowSnapshotImage):

Canonical link: https://commits.webkit.org/282791@main
